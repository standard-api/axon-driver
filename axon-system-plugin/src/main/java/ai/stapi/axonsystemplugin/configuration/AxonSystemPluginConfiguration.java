package ai.stapi.axonsystemplugin.configuration;

import ai.stapi.axonsystem.commandpersisting.CommandMessageStore;
import ai.stapi.axonsystemplugin.DefaultGraphProjection;
import ai.stapi.axonsystemplugin.commandpersisting.CommitCommandFixturesHandler;
import ai.stapi.axonsystemplugin.commandpersisting.CommitCommandFixturesLineRunner;
import ai.stapi.axonsystemplugin.commandpersisting.WipePersistedCommandAfterCommitPolicy;
import ai.stapi.axonsystemplugin.fixtures.FixtureCommandsApplier;
import ai.stapi.axonsystemplugin.fixtures.GenerateFixturesCommandLineRunner;
import ai.stapi.axonsystemplugin.structuredefinition.configure.ConfigureStructureDefinitionHandler;
import ai.stapi.graph.EdgeRepository;
import ai.stapi.graph.NodeRepository;
import ai.stapi.graphoperations.synchronization.GraphSynchronizer;
import ai.stapi.schema.structureSchemaProvider.StructureSchemaFinder;
import ai.stapi.schema.structureSchemaProvider.StructureSchemaProvider;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.axonframework.commandhandling.gateway.CommandGateway;
import org.axonframework.eventhandling.gateway.EventGateway;
import org.springframework.boot.autoconfigure.AutoConfiguration;
import org.springframework.context.ApplicationContext;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.ComponentScan;
import org.springframework.context.annotation.Profile;

@AutoConfiguration
@ComponentScan("ai.stapi.axonsystemplugin.fixtures")
@ComponentScan("ai.stapi.axonsystemplugin.structuredefinition")
public class AxonSystemPluginConfiguration {
  
  @Bean
  public DefaultGraphProjection defaultGraphProjection(
      NodeRepository nodeRepository,
      EdgeRepository edgeRepository,
      GraphSynchronizer graphSynchronizer
  ) {
    return new DefaultGraphProjection(nodeRepository, edgeRepository, graphSynchronizer);
  }
  
  @Bean
  public ConfigureStructureDefinitionHandler configureStructureDefinitionHandler(
      EventGateway eventGateway,
      StructureSchemaProvider structureSchemaProvider,
      StructureSchemaFinder structureSchemaFinder
  ) {
    return new ConfigureStructureDefinitionHandler(
        eventGateway,
        structureSchemaProvider,
        structureSchemaFinder
    );
  }
  
  @Bean
  public FixtureCommandsApplier fixtureCommandsApplier(CommandGateway commandGateway) {
    return new FixtureCommandsApplier(commandGateway);
  }

  @Bean
  public CommitCommandFixturesHandler commitCommandFixturesHandler(
      CommandMessageStore commandMessageStore,
      EventGateway eventGateway,
      ObjectMapper objectMapper
  ) {
    return new CommitCommandFixturesHandler(
        commandMessageStore,
        eventGateway,
        objectMapper
    );
  }
  
  @Bean
  public WipePersistedCommandAfterCommitPolicy wipePersistedCommandAfterCommitPolicy(
      CommandMessageStore commandMessageStore
  ) {
    return new WipePersistedCommandAfterCommitPolicy(commandMessageStore);
  }
  
  @Bean
  @Profile("generate-fixtures")
  public GenerateFixturesCommandLineRunner generateFixturesCommandLineRunner(
      CommandGateway commandGateway,
      DefaultGraphProjection graphProjection,
      ApplicationContext applicationContext
  ) {
    return new GenerateFixturesCommandLineRunner(
        commandGateway,
        graphProjection,
        applicationContext
    );
  }
  
  @Bean
  @Profile("commit-command-fixtures")
  public CommitCommandFixturesLineRunner commitCommandFixturesLineRunner(
      CommandGateway commandGateway,
      ApplicationContext applicationContext
  ) {
    return new CommitCommandFixturesLineRunner(commandGateway, applicationContext);
  }
}
