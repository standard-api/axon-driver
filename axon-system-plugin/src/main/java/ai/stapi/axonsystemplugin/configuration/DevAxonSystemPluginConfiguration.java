package ai.stapi.axonsystemplugin.configuration;

import ai.stapi.axonsystemplugin.aggregatedefinition.CreateAggregateDefinitionPolicy;
import ai.stapi.axonsystemplugin.aggregatedefinition.createCRUDCommandHandlers.CreateAddItemCommandHandlerPolicy;
import ai.stapi.axonsystemplugin.aggregatedefinition.createCRUDCommandHandlers.CreateCreationalCommandHandlerPolicy;
import ai.stapi.axonsystemplugin.aggregatedefinition.createCRUDCommandHandlers.CreateUpdateCommandHandlerPolicy;
import ai.stapi.axonsystemplugin.structuredefinition.configure.ConfigureImportedStructureDefinitionPolicy;
import ai.stapi.graphoperations.graphLoader.inmemory.InMemoryGraphLoaderProvider;
import ai.stapi.graphsystem.aggregatedefinition.model.ResourceAggregateDefinitionMapper;
import ai.stapi.graphsystem.aggregatedefinition.model.eventFactory.CreatedOperationEventFactoriesMapper;
import ai.stapi.graphsystem.aggregatedefinition.model.eventFactory.ItemAddedOperationEventFactoriesMapper;
import ai.stapi.graphsystem.aggregatedefinition.model.eventFactory.UpdatedOperationEventFactoriesMapper;
import ai.stapi.graphsystem.operationdefinition.model.resourceStructureTypeOperationsMapper.AddItemOnResourceOperationsMapper;
import ai.stapi.graphsystem.operationdefinition.model.resourceStructureTypeOperationsMapper.CreationalResourceOperationMapper;
import ai.stapi.graphsystem.operationdefinition.model.resourceStructureTypeOperationsMapper.UpdateResourceOperationMapper;
import ai.stapi.schema.structureSchemaMapper.StructureDefinitionToSSMapper;
import ai.stapi.schema.structureSchemaProvider.StructureSchemaFinder;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.axonframework.commandhandling.gateway.CommandGateway;
import org.springframework.boot.autoconfigure.AutoConfiguration;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Profile;

@AutoConfiguration
@Profile("dev")
public class DevAxonSystemPluginConfiguration {

  @Bean
  public ConfigureImportedStructureDefinitionPolicy configureImportedStructureDefinitionPolicy(
      InMemoryGraphLoaderProvider inMemoryGraphLoaderProvider,
      CommandGateway commandGateway
  ) {
    return new ConfigureImportedStructureDefinitionPolicy(
        commandGateway,
        inMemoryGraphLoaderProvider
    );
  }

  @Bean
  public CreateAggregateDefinitionPolicy createAggregateDefinitionPolicy(
      CommandGateway commandGateway,
      StructureSchemaFinder structureSchemaFinder,
      ResourceAggregateDefinitionMapper aggregateDefinitionMapper,
      ObjectMapper objectMapper
  ) {
    return new CreateAggregateDefinitionPolicy(
        commandGateway,
        structureSchemaFinder,
        aggregateDefinitionMapper,
        objectMapper
    );
  }

  @Bean
  public CreateCreationalCommandHandlerPolicy createCreationalCommandHandlerPolicy(
      CommandGateway commandGateway,
      StructureSchemaFinder structureSchemaFinder,
      ObjectMapper objectMapper,
      StructureDefinitionToSSMapper structureDefinitionToSSMapper,
      CreationalResourceOperationMapper creationalResourceOperationMapper,
      CreatedOperationEventFactoriesMapper createdOperationEventDefinitionMapper
  ) {
    return new CreateCreationalCommandHandlerPolicy(
        commandGateway,
        structureSchemaFinder,
        objectMapper,
        structureDefinitionToSSMapper,
        creationalResourceOperationMapper,
        createdOperationEventDefinitionMapper
    );
  }

  @Bean
  public CreateAddItemCommandHandlerPolicy createAddItemCommandHandlerPolicy(
      CommandGateway commandGateway,
      StructureSchemaFinder structureSchemaFinder,
      ObjectMapper objectMapper,
      StructureDefinitionToSSMapper structureDefinitionToSSMapper,
      AddItemOnResourceOperationsMapper addItemOnResourceOperationsMapper,
      ItemAddedOperationEventFactoriesMapper itemAddedOperationEventMessageMapper
  ) {
    return new CreateAddItemCommandHandlerPolicy(
        commandGateway,
        structureSchemaFinder,
        objectMapper,
        structureDefinitionToSSMapper,
        addItemOnResourceOperationsMapper,
        itemAddedOperationEventMessageMapper
    );
  }

  @Bean
  public CreateUpdateCommandHandlerPolicy createUpdateCommandHandlerPolicy(
      CommandGateway commandGateway,
      StructureSchemaFinder structureSchemaFinder,
      ObjectMapper objectMapper,
      StructureDefinitionToSSMapper structureDefinitionToSSMapper,
      UpdateResourceOperationMapper updateResourceOperationMapper,
      UpdatedOperationEventFactoriesMapper updatedOperationEventFactoriesMapper
  ) {
    return new CreateUpdateCommandHandlerPolicy(
        commandGateway,
        structureSchemaFinder,
        objectMapper,
        structureDefinitionToSSMapper,
        updateResourceOperationMapper,
        updatedOperationEventFactoriesMapper
    );
  }
}
