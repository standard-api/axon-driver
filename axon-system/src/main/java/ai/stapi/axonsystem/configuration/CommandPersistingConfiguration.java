package ai.stapi.axonsystem.configuration;

import ai.stapi.axonsystem.commandpersisting.CommandMessageStore;
import ai.stapi.axonsystem.commandpersisting.InMemoryCommandMessageStore;
import ai.stapi.axonsystem.commandpersisting.PersistCommandInterceptor;
import org.springframework.boot.autoconfigure.AutoConfiguration;
import org.springframework.boot.autoconfigure.condition.ConditionalOnBean;
import org.springframework.boot.autoconfigure.condition.ConditionalOnMissingBean;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Profile;

@AutoConfiguration
public class CommandPersistingConfiguration {
  
  @Bean
  @ConditionalOnMissingBean(CommandMessageStore.class)
  public InMemoryCommandMessageStore inMemoryCommandMessageStore() {
    return new InMemoryCommandMessageStore();
  }
  
  @Bean
  @ConditionalOnBean(CommandMessageStore.class)
  @Profile("dev")
  public PersistCommandInterceptor persistCommandInterceptor(
      CommandMessageStore commandMessageStore
  ) {
    return new PersistCommandInterceptor(commandMessageStore);
  }
}
