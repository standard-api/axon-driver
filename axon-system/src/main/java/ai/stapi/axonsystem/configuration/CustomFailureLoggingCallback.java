package ai.stapi.axonsystem.configuration;

import java.util.logging.Level;
import java.util.logging.Logger;
import javax.annotation.Nonnull;
import org.axonframework.commandhandling.CommandCallback;
import org.axonframework.commandhandling.CommandMessage;
import org.axonframework.commandhandling.CommandResultMessage;

public class CustomFailureLoggingCallback<C, R> implements CommandCallback<C, R> {

  private final CommandCallback<C, R> delegate;
  private final Logger logger;

  public CustomFailureLoggingCallback(Logger logger) {
    this.logger = logger;
    this.delegate = null;
  }

  @Deprecated
  public CustomFailureLoggingCallback(CommandCallback<C, R> delegate) {
    this(
        Logger.getLogger(
            CustomFailureLoggingCallback.class.getSimpleName()
        ),
        delegate
    );
  }

  @Deprecated
  public CustomFailureLoggingCallback(@Nonnull Logger logger,
      @Nonnull CommandCallback<C, R> delegate) {
    this.logger = logger;
    this.delegate = delegate;
  }

  @Override
  public void onResult(
      @Nonnull CommandMessage<? extends C> commandMessage,
      @Nonnull CommandResultMessage<? extends R> commandResultMessage
  ) {
    commandResultMessage.optionalExceptionResult().ifPresent(cause ->
        logger.log(
            Level.WARNING,
            String.format(
                "Command '%s' resulted in %s",
                commandMessage.getCommandName(),
                cause.getClass().getSimpleName()
            ),
            cause
        )
    );
    if (delegate != null) {
      delegate.onResult(commandMessage, commandResultMessage);
    }
  }
}
