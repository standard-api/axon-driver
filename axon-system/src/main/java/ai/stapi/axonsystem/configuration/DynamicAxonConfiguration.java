package ai.stapi.axonsystem.configuration;

import ai.stapi.axonsystem.dynamic.aggregate.AggregateDefinitionDACProvider;
import ai.stapi.axonsystem.dynamic.aggregate.DynamicAggregateConfigurationsProvider;
import ai.stapi.axonsystem.dynamic.aggregate.DynamicAggregateConfigurer;
import ai.stapi.axonsystem.dynamic.aggregate.RuntimeDynamicAggregateConfigurer;
import ai.stapi.axonsystem.dynamic.command.CustomCommandTargetResolver;
import ai.stapi.axonsystem.dynamic.command.DynamicCommandDispatchInterceptor;
import ai.stapi.axonsystem.dynamic.messagehandler.DynamicMessageHandlerLookup;
import ai.stapi.axonsystem.dynamic.messagehandler.DynamicMessageHandlingMemberDefinition;
import ai.stapi.graphsystem.aggregatedefinition.model.AggregateDefinitionProvider;
import ai.stapi.graphsystem.dynamiccommandprocessor.DynamicCommandProcessor;
import org.axonframework.config.Configuration;
import org.axonframework.config.Configurer;
import org.axonframework.eventsourcing.eventstore.EventStore;
import org.axonframework.messaging.annotation.ParameterResolverFactory;
import org.axonframework.modelling.command.CommandTargetResolver;
import org.axonframework.tracing.SpanFactory;
import org.springframework.beans.factory.config.BeanDefinition;
import org.springframework.boot.autoconfigure.AutoConfiguration;
import org.springframework.boot.autoconfigure.condition.ConditionalOnMissingBean;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Lazy;
import org.springframework.context.annotation.Role;

@AutoConfiguration
public class DynamicAxonConfiguration {
  
  @Bean
  public DynamicAggregateConfigurer dynamicAggregateConfigurer(
      DynamicAggregateConfigurationsProvider dynamicAggregateConfigurationsProvider,
      RuntimeDynamicAggregateConfigurer runtimeDynamicAggregateConfigurer
  ) {
    return new DynamicAggregateConfigurer(
        dynamicAggregateConfigurationsProvider,
        runtimeDynamicAggregateConfigurer
    );
  }
  
  @Bean
  @Role(BeanDefinition.ROLE_INFRASTRUCTURE)
  public DynamicMessageHandlerLookup dynamicMessageHandlerLookup() {
    return new DynamicMessageHandlerLookup();
  }
  
  @Bean
  public DynamicMessageHandlingMemberDefinition dynamicMessageHandlingMemberDefinition() {
    return new DynamicMessageHandlingMemberDefinition();
  }
  
  @Bean
  public DynamicCommandDispatchInterceptor dynamicCommandDispatchInterceptor() {
    return new DynamicCommandDispatchInterceptor();
  }
  
  @Bean
  public CustomCommandTargetResolver customCommandTargetResolver() {
    return new CustomCommandTargetResolver();
  }
  
  @Bean
  @ConditionalOnMissingBean(DynamicAggregateConfigurationsProvider.class)
  public AggregateDefinitionDACProvider aggregateDefinitionDACProvider(
      AggregateDefinitionProvider aggregateDefinitionProvider,
      DynamicCommandProcessor dynamicCommandProcessor,
      @Lazy EventStore eventStore,
      SpanFactory spanFactory,
      @Lazy Configuration configuration,
      ParameterResolverFactory parameterResolverFactory,
      CommandTargetResolver commandTargetResolver
  ) {
    return new AggregateDefinitionDACProvider(
        aggregateDefinitionProvider,
        dynamicCommandProcessor,
        eventStore,
        spanFactory,
        configuration,
        parameterResolverFactory,
        commandTargetResolver
    );
  }
  
  @Bean
  public RuntimeDynamicAggregateConfigurer runtimeDynamicAggregateConfigurer(
      DynamicAggregateConfigurationsProvider dynamicAggregateConfigurationsProvider,
      @Lazy Configurer configurer
  ) {
    return new RuntimeDynamicAggregateConfigurer(
        dynamicAggregateConfigurationsProvider,
        configurer
    );
  }
}
