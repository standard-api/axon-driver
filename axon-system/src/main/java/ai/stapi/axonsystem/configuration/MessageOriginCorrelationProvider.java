package ai.stapi.axonsystem.configuration;

import org.axonframework.messaging.correlation.CorrelationDataProvider;
import org.axonframework.messaging.correlation.MessageOriginProvider;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;

@Configuration
public class MessageOriginCorrelationProvider {

  @Bean
  public static CorrelationDataProvider messageCorrelationProvider() {
    return new MessageOriginProvider();
  }
}
