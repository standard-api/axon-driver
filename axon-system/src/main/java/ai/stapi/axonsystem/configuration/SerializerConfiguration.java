package ai.stapi.axonsystem.configuration;

import ai.stapi.configuration.SerializationConfiguration;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.axonframework.serialization.Serializer;
import org.axonframework.serialization.json.JacksonSerializer;
import org.springframework.beans.factory.annotation.Qualifier;
import org.springframework.boot.autoconfigure.AutoConfiguration;
import org.springframework.boot.autoconfigure.AutoConfigureAfter;
import org.springframework.boot.autoconfigure.condition.ConditionalOnBean;
import org.springframework.context.annotation.Bean;

@AutoConfiguration
@AutoConfigureAfter(SerializationConfiguration.class)
public class SerializerConfiguration {
  
  @Bean
  @Qualifier("messageSerializer")
  @ConditionalOnBean(ObjectMapper.class)
  public Serializer customMessageSerializer(ObjectMapper objectMapper) {
    return JacksonSerializer
        .builder()
        .objectMapper(objectMapper)
        .build();
  }
}
