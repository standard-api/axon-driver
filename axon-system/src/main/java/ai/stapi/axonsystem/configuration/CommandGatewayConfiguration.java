package ai.stapi.axonsystem.configuration;

import ai.stapi.axonsystem.commandvalidation.CommandValidatorDispatchInterceptor;
import ai.stapi.axonsystem.configuration.implementations.CommandDispatchedAtInterceptor;
import ai.stapi.axonsystem.configuration.implementations.CustomFailureLoggingCallback;
import ai.stapi.graphsystem.commandvalidation.model.CommandValidator;
import java.util.List;
import java.util.logging.Logger;
import org.axonframework.commandhandling.CommandBus;
import org.axonframework.commandhandling.CommandMessage;
import org.axonframework.commandhandling.DuplicateCommandHandlerResolver;
import org.axonframework.commandhandling.SimpleCommandBus;
import org.axonframework.commandhandling.gateway.CommandGateway;
import org.axonframework.commandhandling.gateway.DefaultCommandGateway;
import org.axonframework.common.transaction.TransactionManager;
import org.axonframework.messaging.MessageDispatchInterceptor;
import org.axonframework.messaging.MessageHandlerInterceptor;
import org.axonframework.messaging.correlation.CorrelationDataProvider;
import org.axonframework.messaging.correlation.MessageOriginProvider;
import org.axonframework.messaging.interceptors.CorrelationDataInterceptor;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.autoconfigure.AutoConfiguration;
import org.springframework.boot.autoconfigure.condition.ConditionalOnMissingBean;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Profile;

@AutoConfiguration
public class CommandGatewayConfiguration {

  @Bean
  @ConditionalOnMissingBean(CommandGateway.class)
  public DefaultCommandGateway configuredCommandGateway(
      @Autowired CommandBus commandBus,
      @Autowired
      List<MessageDispatchInterceptor<? super CommandMessage<?>>> messageDispatchInterceptors
  ) {
    var failureCommandCallback = new CustomFailureLoggingCallback<>(
        Logger.getLogger(DefaultCommandGateway.class.getSimpleName())
    );
    return DefaultCommandGateway.builder()
        .commandBus(commandBus)
        .commandCallback(failureCommandCallback)
        .dispatchInterceptors(messageDispatchInterceptors)
        .build();
  }

  @Bean
  @ConditionalOnMissingBean
  public CommandBus configuredCommandBus(
      @Autowired TransactionManager txManager,
      @Autowired org.axonframework.config.Configuration axonConfiguration,
      @Autowired DuplicateCommandHandlerResolver duplicateCommandHandlerResolver,
      @Autowired
      List<MessageHandlerInterceptor<? super CommandMessage<?>>> messageHandlerInterceptors
  ) {
    var commandBus = SimpleCommandBus.builder()
        .transactionManager(txManager)
        .duplicateCommandHandlerResolver(duplicateCommandHandlerResolver)
        .spanFactory(axonConfiguration.spanFactory())
        .messageMonitor(axonConfiguration.messageMonitor(CommandBus.class, "commandBus"))
        .build();

    commandBus.registerHandlerInterceptor(
        new CorrelationDataInterceptor<>(axonConfiguration.correlationDataProviders())
    );

    messageHandlerInterceptors.forEach(commandBus::registerHandlerInterceptor);

    return commandBus;
  }
  
  @Bean
  public CommandDispatchedAtInterceptor commandDispatchedAtInterceptor() {
    return new CommandDispatchedAtInterceptor();
  }
  
  @Bean
  @Profile("dev")
  public CommandValidatorDispatchInterceptor commandValidatorDispatchInterceptor(
      CommandValidator commandValidator
  ) {
    return new CommandValidatorDispatchInterceptor(commandValidator);
  }

  @Bean
  public CorrelationDataProvider messageCorrelationProvider() {
    return new MessageOriginProvider();
  }
}
