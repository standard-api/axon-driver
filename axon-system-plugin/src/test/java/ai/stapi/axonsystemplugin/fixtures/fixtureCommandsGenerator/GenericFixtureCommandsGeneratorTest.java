package ai.stapi.axonsystemplugin.fixtures.fixtureCommandsGenerator;

import ai.stapi.axonsystemplugin.exampleimplmentation.configuration.ExampleModuleDefinitionsLoader;
import ai.stapi.axonsystemplugin.fixtures.ExampleImplementations.ExampleAnotherFixtureCommandsGenerator;
import ai.stapi.axonsystemplugin.fixtures.ExampleImplementations.ExampleFileFixtureCommandsGenerator;
import ai.stapi.axonsystemplugin.fixtures.ExampleImplementations.ExampleFixtureCommandsGenerator;
import ai.stapi.axonsystemplugin.fixtures.ExampleImplementations.ExampleFixtureTags;
import ai.stapi.graphsystem.fixtures.fixtureCommandsGenerator.FixtureCommandsGenerator;
import ai.stapi.graphsystem.fixtures.fixtureCommandsGenerator.GenericFixtureCommandsGenerator;
import ai.stapi.schema.adHocLoaders.FileLoader;
import ai.stapi.test.domain.DomainTestCase;
import ai.stapi.test.schemaintegration.StructureDefinitionScope;
import com.fasterxml.jackson.databind.ObjectMapper;
import java.util.List;
import java.util.Set;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;

@StructureDefinitionScope(ExampleModuleDefinitionsLoader.SCOPE)
public class GenericFixtureCommandsGeneratorTest extends DomainTestCase {

  @Autowired
  private ObjectMapper objectMapper;
  @Autowired
  private FileLoader fileFixtureLoader;

  @Test
  void itShouldReturnNoCommandsWhenNoFixtureTagProvided() {
    var exampleGenerator = new ExampleFixtureCommandsGenerator();
    var genericGenerator = new GenericFixtureCommandsGenerator(
        List.of(exampleGenerator)
    );
    var actualResults = genericGenerator.generate(
        List.of(),
        List.of(),
        Set.of(),
        0,
        FixtureCommandsGenerator.PRIORITY_STRUCTURES
    ).toList();
    this.thenObjectApprovedWithShownIds(actualResults);
  }

  @Test
  void itShouldReturnNoCommandsWhenUnusedFixtureTagProvided() {
    var exampleGenerator = new ExampleFixtureCommandsGenerator();
    var genericGenerator = new GenericFixtureCommandsGenerator(
        List.of(exampleGenerator)
    );
    var providedTags = List.of(ExampleFixtureTags.EXAMPLE_UNUSED_TAG);
    var actualResults = genericGenerator.generate(
        providedTags,
        List.of(),
        Set.of(),
        0,
        FixtureCommandsGenerator.PRIORITY_STRUCTURES
    ).toList();
    this.thenObjectApprovedWithShownIds(actualResults);
  }

  @Test
  void itShouldReturnCorrectCommandsWhenExampleFixtureTagProvided() {
    var exampleGenerator = new ExampleFixtureCommandsGenerator();
    var genericGenerator = new GenericFixtureCommandsGenerator(
        List.of(exampleGenerator)
    );
    var actualResults = genericGenerator.generate(
        List.of(ExampleFixtureTags.EXAMPLE_TAG),
        List.of(),
        Set.of(),
        0,
        FixtureCommandsGenerator.PRIORITY_STRUCTURES
    ).toList();
    this.thenObjectApprovedWithShownIds(actualResults);
  }

  @Test
  void itShouldLoadFixturesFromFile() {
    var exampleGenerator =
        new ExampleFileFixtureCommandsGenerator(this.objectMapper, fileFixtureLoader);
    var genericGenerator = new GenericFixtureCommandsGenerator(
        List.of(exampleGenerator)
    );
    var actualResults = genericGenerator.generate(
        List.of(
            ExampleFixtureTags.EXAMPLE_FILE_LOADER_TAG
        ),
        List.of(),
        Set.of(),
        0,
        FixtureCommandsGenerator.PRIORITY_STRUCTURES
    ).toList();
    this.thenObjectApprovedWithShownIds(actualResults);
  }

  @Test
  void itShouldLoadOnlyFixturesFromFileWhichHaveNotBeenExecutedYet() {
    var exampleGenerator =
        new ExampleFileFixtureCommandsGenerator(this.objectMapper, fileFixtureLoader);
    var genericGenerator = new GenericFixtureCommandsGenerator(
        List.of(exampleGenerator)
    );
    var actualResults = genericGenerator.generate(
        List.of(
            ExampleFixtureTags.EXAMPLE_FILE_LOADER_TAG
        ),
        List.of(),
        Set.of("CreateStructureDefinition/230223011314.example.profile.canonical.json"),
        0,
        FixtureCommandsGenerator.PRIORITY_STRUCTURES
    ).toList();
    this.thenObjectApprovedWithShownIds(actualResults);
  }

  @Test
  void itShouldReturnCorrectCommandsOnlyOnceWhenMoreTagsSupportedByOneGeneratorProvided() {
    var exampleGenerator = new ExampleFixtureCommandsGenerator();
    var genericGenerator = new GenericFixtureCommandsGenerator(
        List.of(exampleGenerator)
    );
    var providedTags =
        List.of(ExampleFixtureTags.EXAMPLE_TAG, ExampleFixtureTags.EXAMPLE_ANOTHER_TAG);
    var actualResults = genericGenerator.generate(
        providedTags,
        List.of(),
        Set.of(),
        0,
        FixtureCommandsGenerator.PRIORITY_STRUCTURES
    ).toList();
    this.thenObjectApprovedWithShownIds(actualResults);
  }

  @Test
  void itShouldReturnCorrectCommandsWhenTagIsSupportedByMoreGenerators() {
    var exampleGenerator = new ExampleFixtureCommandsGenerator();
    var exampleGenerator2 = new ExampleAnotherFixtureCommandsGenerator();
    var genericGenerator = new GenericFixtureCommandsGenerator(
        List.of(
            exampleGenerator,
            exampleGenerator2
        )
    );
    var providedTags = List.of(ExampleFixtureTags.EXAMPLE_ANOTHER_TAG);
    var actualResults = genericGenerator.generate(
        providedTags,
        List.of(),
        Set.of(),
        0,
        FixtureCommandsGenerator.PRIORITY_STRUCTURES
    ).toList();
    this.thenObjectApprovedWithShownIds(actualResults);
  }
}
