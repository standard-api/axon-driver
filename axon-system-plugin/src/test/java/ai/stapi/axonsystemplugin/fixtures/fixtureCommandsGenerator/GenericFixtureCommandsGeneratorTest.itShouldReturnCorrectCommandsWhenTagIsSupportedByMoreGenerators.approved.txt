[ {
  "generatorClassName" : "ai.stapi.axonsystemplugin.fixtures.ExampleImplementations.ExampleFixtureCommandsGenerator",
  "oneTimeGenerator" : true,
  "commandDefinitions" : [ {
    "name" : "ai.stapi.axonsystemplugin.exampleimplmentation.CreateNewExampleNodeCommand",
    "command" : {
      "serializationType" : "CreateNewExampleNodeCommand",
      "targetIdentifier" : {
        "id" : "2d82aa9a-64d9-4e68-a825-ecacca941201"
      },
      "exampleNodeName" : "Example Name",
      "exampleNodeType" : "ExampleNodeType",
      "exampleQuantity" : 0,
      "exampleStringAttribute" : "default",
      "exampleListAttribute" : [ ]
    }
  } ],
  "processedFiles" : [ ]
}, {
  "generatorClassName" : "ai.stapi.axonsystemplugin.fixtures.ExampleImplementations.ExampleAnotherFixtureCommandsGenerator",
  "oneTimeGenerator" : true,
  "commandDefinitions" : [ {
    "name" : "ai.stapi.axonsystemplugin.exampleimplmentation.CreateNewExampleNodeCommand",
    "command" : {
      "serializationType" : "CreateNewExampleNodeCommand",
      "targetIdentifier" : {
        "id" : "2d82aa9a-64d9-4e68-a825-ecacca941201"
      },
      "exampleNodeName" : "Example Another Name",
      "exampleNodeType" : "ExampleNodeType",
      "exampleQuantity" : 0,
      "exampleStringAttribute" : "default",
      "exampleListAttribute" : [ ]
    }
  } ],
  "processedFiles" : [ ]
} ]